### ISSUE TITLE
Simple mappings fail

### ISSUE THREAD

**Issue by GoogleCodeExporter** [vadbch/orika#90]
```
I am trying to map type A to C, then B to C. I could reproduce with really 
simple classes, A, B, C being a class with a single string attribute named "s".

This fails with exception "B is an unsupported source class". Looks like Orika 
uses the first converter: it expects A and gets B instead.

The exception goes away when I explicitly register the mappings A->C, B->C. 
This is tedious because this forces me to register all possible mappings (a lot 
in my case!) during initialization.

Please check out attached test case.

I could reproduce the issue with Orika 1.4.1 and 1.4.2-SNAPSHOT.
```

Original issue reported on code.google.com by `oche...@gmail.com` on 15 Mar 2013 at 11:59

Attachments:
- [TestCase.java](https://storage.googleapis.com/google-code-attachments/orika/issue-90/comment-0/TestCase.java)

**Comment by GoogleCodeExporter** [comment]
```




```

Original comment by `elaat...@gmail.com` on 15 Mar 2013 at 2:20
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
Looks to me like the 
ma.glasnost.orika.impl.DefaultMapperFactory.lookupObjectFactory(Type<T>) needs 
to accept a source as well as a destination for its signature.  That could mean 
a lot more source generation every time a new source type was found though.  
The quick and dirty solution is to to provide your own DefaultMapperFactory 
overriding:

    @Override
    public <T> ObjectFactory<T> lookupObjectFactory(Type<T> targetType) {
        return new DefaultConstructorObjectFactory<T>(targetType.getRawType());
    }

This means you'll never end up with specific source types in object factory 
generated code.  Obviously, there's a performance hit with this... so the 
options are:
1. Register your classes (maybe automate this?) for best performance.
2. Orika regenerates code every time it finds another source type for good 
amortised performance.
3. Create your own mapper factory as above for sub-optimal performance, but you 
can do it now (note this also involves some code smelling as you will need to 
recreate ConfigurableMapper).

Monsieur elaatifi - your work is most appreciated!
```

Original comment by `damnpana...@gmail.com` on 21 Mar 2013 at 10:36

**Comment by GoogleCodeExporter** [comment]
```
Thank you Dean, for your suggestions, as you said pre-register classes is the 
best solution.

We want to let this issue Open for further discussions, suggestions. Leverage 
one ObjectFactory per source and destination classes may hit performance and 
need to change the API.
```

Original comment by `elaat...@gmail.com` on 22 Mar 2013 at 11:00

**Comment by GoogleCodeExporter** [comment]
```
Take a look at the current master; an initial implementation of this behavior 
is available. The api has been extended for MapperFactory to add new signatures 
for lookupObjectFactory and registerObjectFactory and deprecate the old ones, 
without any change to ObjectFactory interface.
(This seems to have resolved the test case in this issue.)
```

Original comment by `matt.deb...@gmail.com` on 30 May 2013 at 7:38
- Changed state: **Started**

**Comment by GoogleCodeExporter** [comment]
```
Fixed in 1.4.3 release
```

Original comment by `matt.deb...@gmail.com` on 28 Jun 2013 at 8:20
- Changed state: **Fixed**
